### PR TITLE
[CDAP-9207] Multi search filter

### PR DESCRIPTION
**Pull request by rohitcmu**
https://issues.cask.co/browse/CDAP-9207

### CODE REVIEW

**Comment by elfenheart** [comment]
@rohitcmu Why do we need to add a new dependency for angular filter? Is there something that the default `$filter` cannot do?

File: cdap-ui/app/filters/multi-key-search.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Filter that searches through multiple keys of an object for the search text.
+ * This does a case insensitive "contains" match for the search term.
+ * @param {Array} input Items to search through.
+ * @param {Array} keys Keys to match.
+ * @param {String} search Search terms.
+ * @return {Array} filtered items.
+ */
+angular.module(PKG.name+'.filters').filter('myMultiKeySearch', function() {
+  return function multiKeySearch(input, keys, search) {
+    if (!angular.isArray(keys) || !search) {
```
Comment (elfenheart) [inline cdap-ui/app/filters/multi-key-search.js]: might have to check for the length of `keys` as well

File: cdap-ui/app/filters/multi-key-search.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      return input;
+    }
+    search = search.toLowerCase();
+    var filtered = input.filter(function (value) {
```
Comment (ajainarayanan) [inline cdap-ui/app/filters/multi-key-search.js]: Could be simplified to,

```
return input.filter(value =>
  keys.filter(key => value[key] && value[key].toLowerCase().indexOf(search) !== -1).length
);
```

**Review by ajainarayanan — COMMENTED** [review]
Minor comment. Otherwise 👍 LGTM Ready for review by Lea or @bdmogal

**Review by ajainarayanan — APPROVED** [review]
👍 LGTM Ready for review from Lea and @bdmogal

**Comment by leacd** [comment]
I have a few points about this feature: 
- I was a little confused by the implementation. It is a search box, but has a "filter" label on it. Usually, filters are an iterative faceted search which gives the user pre-set criteria, while a search is broader and requires the user to input queries.  The implementation seems to lean toward this feature being a search, although we allow the user to only search using very specific queries. I want to suggest to change the label to describe the search and its parameters: e.i. "Search plugin names and type" please add a search icon (magnifying glass)
- I was also confused by the behavior: the search/filter doesn't work across all plugins, rather it filter/searches only the section that is currently open. Because  Can we change it to search/filter across all sections?

**Comment by ajainarayanan** [comment]
@leacd Just so we are on the same page, 
1. The filter is represented as filter. Not sure where we gave an impression that it was a search.
This is what I am seeing 
screen shot 2017-04-04 at 7 29 45 am

To me when I see filter I assume its going to filter by whatever I am typing. If we are changing the filter/search being done in the left panel then it might be outside the scope of this PR. Lets file a JIRA and add the changes as part of a separate PR. 

2. To show only the current section opened is a decision that we took. We initially opened all the sections while searching however when the no of sections increased it didn't make sense to open all the sections. Even though we don't open all the sections if you open any closed section it should only show the plugins that match the filter.
filter

**Comment by leacd** [comment]
Not withstanding the comments above. LGTM.

**Pull request merged**